File: frontend/src/types/bucket.ts
export interface Bucket {
  bucketId: string;
  name: string;
  description: string;
  tags: string[];
  userId: string;
  sourceIds?: string[]; //list of misc source ids
  imageKeys: string[];
  created: string;
  updated: string;
  visibility: "Private" | "Public" | "Invite";
  likes: string[];
  iterations: string[];
  iteratedFrom?: string;
}

export interface UpdateBucket {
  name: string;
  description: string;
  visibility: "Private" | "Public" | "Invite";
  tags?: string[];
}

export type BucketTag = {
  value: string;
  label: string;
};

export const tagsList: BucketTag[] = [
  {
    value: "Research",
    label: "research",
  },
  {
    value: "AI",
    label: "ai",
  },
  {
    value: "Technology",
    label: "technology",
  },
  {
    value: "Science",
    label: "science",
  },
  {
    value: "Politics",
    label: "politics",
  },
  {
    value: "Business",
    label: "business",
  },
  {
    value: "Economy",
    label: "economy",
  },
  {
    value: "Food",
    label: "food",
  },
  {
    value: "Travel",
    label: "travel",
  },
  {
    value: "Health",
    label: "health",
  },
  {
    value: "Entertainment",
    label: "entertainment",
  },
  {
    value: "Sports",
    label: "sports",
  },
  {
    value: "World",
    label: "world",
  },
  {
    value: "Other",
    label: "other",
  },
];

File: backend/src/models/bucket.py
from pydantic import BaseModel
from typing import List, Optional, Literal
from datetime import datetime
from src.db.mongodb import get_collection

Buckets = get_collection("buckets")


class Bucket(BaseModel):
    bucketId: str
    name: str
    description: str
    tags: list[str]
    userId: str
    articleIds: list[str]  # list of article ids to fetch from neo4j
    sourceIds: list[str]
    imageKeys: list[str]
    created: datetime
    updated: datetime
    visibility: Literal["Private", "Public", "Invite"]
    likes: list[str]
    iteratedFrom: Optional[str]  # userId
    iterations: list[str]
    imageKeys: Optional[list[str]]


class BucketConfig(BaseModel):
    name: str
    description: str
    visibility: Literal["Private", "Public", "Invite"]
    tags: list[str] = []
    sourceIds: list[str] = []
    imageKeys: list[str] = []


class UpdateBucket(BaseModel):
    name: Optional[str]
    description: Optional[str]
    visibility: Literal["Private", "Public", "Invite"]


class IterateBucket(BaseModel):
    name: str
    description: str


class LikeBucket(BaseModel):
    bucketId: str


class UnLikeBucket(BaseModel):
    bucketId: str
